webhook/workload/mutating/workload_update_handler.go

```diff
@@ -0,0 +1,253 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mutating
+
+import (
+	"context"
+	"encoding/json"
+	"net/http"
+	"reflect"
+
+	kruiseappsv1alpha1 "github.com/openkruise/kruise-api/apps/v1alpha1"
+	appsv1alpha1 "github.com/openkruise/rollouts/api/v1alpha1"
+	"github.com/openkruise/rollouts/pkg/util"
+	admissionv1 "k8s.io/api/admission/v1"
+	apps "k8s.io/api/apps/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// WorkloadHandler handles Pod
+type WorkloadHandler struct {
+	// To use the client, you need to do the following:
+	// - uncomment it
+	// - import sigs.k8s.io/controller-runtime/pkg/client
+	// - uncomment the InjectClient method at the bottom of this file.
+	Client client.Client
+
+	// Decoder decodes objects
+	Decoder *admission.Decoder
+	Finder  *util.ControllerFinder
+}
+
+var _ admission.Handler = &WorkloadHandler{}
+
+// Handle handles admission requests.
+//  TODO
+//  Currently there is an implicit condition for rollout: the workload must be currently in a stable version (only one version of Pods),
+//  if not, it will not enter the rollout process. There is an additional problem here, the user may not be aware of this.
+//  when user does a release and thinks it enters the rollout process, but due to the implicit condition above,
+//  it actually goes through the normal release process. No good idea to solve this problem has been found yet.
+func (h *WorkloadHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	// if subResources, then ignore
+	if req.Operation != admissionv1.Update || req.SubResource != "" {
+		return admission.Allowed("")
+	}
+
+	switch req.Kind.Group {
+	// kruise cloneSet
+	case kruiseappsv1alpha1.GroupVersion.Group:
+		if req.Kind.Kind != util.ControllerKruiseKindCS.Kind {
+			return admission.Allowed("")
+		}
+		// check deployment
+		newObj := &kruiseappsv1alpha1.CloneSet{}
+		if err := h.Decoder.Decode(req, newObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		copy := newObj.DeepCopy()
+		oldObj := &kruiseappsv1alpha1.CloneSet{}
+		if err := h.Decoder.Decode(
+			admission.Request{AdmissionRequest: admissionv1.AdmissionRequest{Object: req.AdmissionRequest.OldObject}},
+			oldObj); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		if err := h.handlerCloneSet(newObj, oldObj); err != nil {
```

Comment:
handlerCloneSet can add an extra return value called "changed", so that we can avoid unnecessary heavy deepequal operation below

Reply:
done